### ISSUE TITLE
Invidiual module pre-compilation API

### ISSUE THREAD

**Issue by tonytanger** [systemjs/builder#226]
I'm working on a project that requires me to use `csp` version of `system.js`, so I have to send to the client pre-compiled files. Another requirement is that I'm trying to optimize for `HTTP/2`, so I would like to compile each module/file individually and not bundle them. ESM can be done easily with babel or even with `builder.trace(filename)`, but CommonJS files are not as easy.

I've tried to use `builder.trace(filename)` but what it returned is not compiled to system. I tried `builder.build(filename)` but it bundles the files.

What's the best way to pre-compiled CommonJS files individually, and load them into the browser?

Thanks in advance for your help!

**Comment by guybedford** [comment]
Yes we are missing an API here to hook into individual file compilations. It is definitely needed for these types of use cases.

**Comment by guybedford** [comment]
Ideally this API should work in such a way that it doesn't require the full trace either, but can act on a single arbitrary module:

```javascript
builder.compileModule(moduleName, cache)
```

**Comment by guybedford** [comment]
The individual compile API is actually just:

```javascript
builder.build('[moduleName]')
```

That should be enough actually!

**Comment by guybedford** [comment]
If the API we stick with for this is `builder.build('[moduleName]')` then we need to optimize this arithmetic expression to not need to trace the whole tree. Tracking this optimization here.

//cc @Munter

**Comment by Munter** [comment]
I did a bit of work on this yesterday and ran into the exact problem you already mentioned here. The builder calls `System.import` which ends up tracing the whole tree. My thinking is that all the resolving of file names at the time of request to the server would already be handled though, so the question is, can I construct a different path that just reads the file from disk and stuffs it directly into the transpiler chain, or do we need to figure out a mode where SystemJS itself doesn't trace the complete tree?

**Comment by guybedford** [comment]
If you can manually construct a `load` record just like the traced load records, you could pass that directly to the private API - https://github.com/systemjs/builder/blob/master/lib/compile.js#L43. Ideally we would have this optimization in the arithmetic though.

**Comment by Munter** [comment]
I'm super frustrated with this at the moment. I can't figure out the control flow between the builder and systemjs, so I can't see what happens or where I can hook in. I can't get the sources correctly linked up so I can even get console debugging in the systemjs part of the code.

I could maybe construct a `load` record by hand if I knew what plugins to run, but that information is not available  to me if I don't use systemjs to resolve things for me.

You're saying we should have the `[module]` arithmetic expression not trace the whole tree, but from what I can gather the tracing doesn't happen in the arithmetic at all, but is offloaded to systemjs and the information about only needing the first level is lost in the process.

I don't really know how to proceed

**Comment by guybedford** [comment]
Yes we use SystemJS itself to do the trace which is a non-spec API in the module loader. This works well in ensuring we have the exact same code doing module tracing client and server, but does make it difficult to intercept the process.

That's why I was suggesting skipping that entirely and constructing a custom load record, but yes then it means having to work out all the information to associate with that manually, which is a complex boundary. If it's not linking up I wouldn't suggest wasting any more time on it.

Will see what I can do on this tomorrow.

**Comment by guybedford** [comment]
This has been fixed in https://github.com/systemjs/builder/commit/42bf392606dc6c8e1ede859c9a6ad74e6a099b50 where `builder.trace('[moduleName]')` will only now read that one module.

Along with that work, I've also added the trace cache now, accessible via `builder.cache.trace`, where it can be both read and populated by canonical module name keys.

@Munter hopefully this allows you to pick up on this, just let me know if you have any questions at all.

**Comment by Munter** [comment]
Awesome. I'll start playing with it now and will report back.

**Comment by guybedford** [comment]
Great, let me know how it goes.

**Comment by Munter** [comment]
Ok, I have a few things working now, but have run into some issues. The repository can be found here: https://github.com/Munter/express-systemjs-translate

Even though I instantiate the builder with a `baseURL` that points at my asset folder, it seems SystemJS behind it is not getting and/or using that `baseURL`. It seems to be using `process.cwd()`. So I had to do this: https://github.com/Munter/express-systemjs-translate/blob/master/lib/index.js#L95
Why does SystemJS ignore the `baseURL` setting? Also, I think I might need to mess more with the paths here to guarantee that things work in all cases.

When I compile a javascript asset with a syntax error I expect te compiler to throw and I fall back to sending the uncompiled asset over the wire to give the consumer a chance to debug it on their end with browser tools. This happens here: https://github.com/Munter/express-systemjs-translate/blob/master/lib/index.js#L117-L121
However for some reason I need to explicitly handle the rejection [on line 110](https://github.com/Munter/express-systemjs-translate/blob/master/lib/index.js#L110) and pass it on to get the right behavior. Why?

In my test where I try to load jquery the builder traces the entire tree (a problem handled in a different issue), but when resolving urls for dependencies it again seems like the `baseURL` is ignored, making it fail.

The good news is that it is sort of working :)

**Comment by guybedford** [comment]
Awesome!!

If req.path starts with `/` then that may be seen as `System.import('/path/here')` which would on a filesystem become `file:///path/here`? Perhaps make sure to resolve paths relative to the "public server folder" on the server (a different concept to baseURL and paths, which are shared modular reference points) - `path.resolve(publicPath, req.path)`, before importing `System.import('/path/to/public/server/url/path/here')` which will work as we use fully-qualified URLs in both environments. Note also the following function may be useful at some point for converting from these URLs into unique canonical names that can be imported more portibly (https://github.com/systemjs/builder/blob/master/lib/builder.js#L61).

With the errors, I'm wondering if this might be some sort of Promise compatibility issue. SystemJS will use whatever Promise is present globally, loading when.js only when needed. Perhaps try using `var Promise = require('when/es6-shim/Promise')` - that might affect compatibility?

I was thinking about the Accept header again, and we should probably just check for `application/javascript` and return `System.register` output for that header regardless. The reason being that for CSP environments we use `<script>` tag injection which will only ever have that accept header. For scripts that users want to always load with normal `<script>` tags and not through SystemJS, users can use meta configuration to declare them as `global` scripts, and then this would work naturally provided that for single-file compiles we don't compile `global` or `amd` scripts into `System.registerDynamic`. I've created an issue to track this at https://github.com/systemjs/builder/issues/247.

A bit of a mouthful I know, but let me know if I can clarify any of the above.

**Comment by Munter** [comment]
I guess the most important path to get right is the one where I inject the compiler result into the SystemJS cache in the response. The others I can test for. Could you please look at this expectation and tell me if the path injected into the cache is correct given that `baseURL` in the systemjs runtime is `fixtures`.
What does SystemJS use as a cache key?

**Comment by Munter** [comment]
Triggering a middleware intercept on any `application/javascript` request will yield many false positives. Any script tag would be intercepted, including the load of `system.js` itself. Maybe a GET param is needed afterall

**Comment by guybedford** [comment]
@Munter this is why I'm suggesting maintaining an equality transform for plain global scripts in https://github.com/systemjs/builder/issues/247 where possible. Such that a transform of SystemJS itself through the builder would return the script unmodified. The issue arises if a normal global is detected as CommonJS or ES module (which is an altering transform) instead of global or AMD (which can be made non-altering here). ES module detection is not usually mistaken though, which just leaves the over-classification of CommonJS modules. It feels like an edge that can be handled though.

Basically, I think we must put the CSP use case first - which doesn't allow a GET param or any adjustments. It's just a `<script>` tag injection. This is the potential HTTP/2 production workflow in future, so is important to ensure it is given the necessary support.

**Comment by guybedford** [comment]
It's a tough problem though - not pretending to know the answers, just trying to work through the space.

**Comment by guybedford** [comment]
@Munter sorry misread here, you mean a GET param don't you? `/module/name.js?module` or something like that? That might well work here.

**Comment by guybedford** [comment]
(such a direction would be made easier by a `System.queryParam` option or similar, so let me know)

**Comment by guybedford** [comment]
This is how I would approach the problem:

* Only modify sources contained within the baseURL or wildcard paths entries.
* Users that don't want sources modified can make sure such scripts are not within the baseURL or wildcard paths entries.

This avoids the need for any _branching_ and would make the server effectively behave the same as the jspm CDN is set to work within those _module paths_.

**Comment by Munter** [comment]
`system.js` is inside `baseURL`

**Comment by Munter** [comment]
I looks like `path/to/module.js?ThisIsAModule` is the way to go

**Comment by guybedford** [comment]
This work ties in quite closely to where the jspm CDN is headed, so it may help to clarify a little in terms of what the current plans are:
* The idea is for all resources accessible on the CDN currently to run through this exact same style of single-file compilation. JS files and plugin resources will be converted into System.registerDynamic or System.register output.
* This gives the optimum HTTP/2 CSP-compatible caching workflow where even .json file loads come with CSP-support.
* This does mean we've altered the original sources, so that expectations about file contents are lost. But I don't see that as an issue at all. This is a module system talking to a module server. The communication happens to take place over HTTP but it should be in the native language of modules. The only risk is if a user expects to load a `<link>` or `<script>` tag to something hard-coding in a module path (which by their nature discourage this that they are not simple names).
* Only resources configured to be loaded via a plugin will be compiled (`'*.jsx': { loader: 'jsx' }`) so that this is opt-in and not desctructive. The last case left is then just global scripts that the user wants to be `<script>` tag accessible, which was the reason for creating https://github.com/systemjs/builder/issues/247 to enable this when the `format` and `exports` are explicitly set for such a compatibility.

Do you agree that the benefits of the above in having fully-cached module URLs far outweigh the costs that JS and some resources might not always look exactly as expected? In the same way that minification is a server-level optimization, this type of rewriting is part of that.

If you can agree with me on the above, then for JSPM itself, it will still continue to download and provide the user folders which contain the **uncompiled resources**. That is the `jspm_packages/registry/...` folders will contain raw JSX and ES6 modules. This is the development workflow that allows a user to make a quick edit to any package anywhere in their app and get an immediate reflection of that change, giving them full control over the original source code.

In such a future, it would be awesome for there to be a `jspm serve` production / development server. This would then turn the local `jspm_packages` and `app` package containing my own code (configured via the package.json) to make them serve compiled sources directly to the browser with a caching layer. That is the `directories.lib` and `jspm_packages/registry/*` **module paths** folders specifically would be compiled. Everything else would continue to be served normally to retain backwards-compatibility with normal script tag behaviours. This gives the user a cacheable dev / production server with HTTP/2 & CSP support in a way that is identical to the CDN behaviour - they can switch between the two seamlessly and there are no implementation differences. That `system.js` is within `jspm_packages` is an exception to the rule above.

In terms of a SystemJS development server today, perhaps a better way of phrasing the above is that only certain URL patterns should be regarded as "module paths". Typically this will just be my application code folder and jspm_packages or vendor subfolders. Bower workflows obviously make this more complex, but the principle of URL-based selection for module serving over loader-level requests of a module remains the same.

Let me know your thoughts, it's great to be able to discuss these ideas.

**Comment by guybedford** [comment]
When ES modules are finally supported in browsers, the transpilation output would no longer be `System.register` but plain ES modules. Such a change most likely would hinge on the Accept header in due course, which would implement without CSP issues in the real module loader due to it not needing `<script>` injection.

**Comment by guybedford** [comment]
This issue is still being tracked in https://github.com/systemjs/builder/issues/256.
